States service impl test

gabichelsea opened the pull request:
Trello card link:

[https://trello.com/c/kH0c2HSu](https://trello.com/c/kH0c2HSu)

gabichelsea committed:
Refactor the getCandidateApplicationsByCandidateId() method name
Done validation cases, empty list and a list with single element cases for the getStatesByApplicationId() method

gabichelsea committed:
Refactor

gabichelsea committed:
Check state view list by creation date

gabichelsea committed:
Introduce setUp() method with @Before annotation

gabichelsea committed:
Add validation
Add test cases for saveState() method
Replace is() to equalTo() if it is better

gabichelsea committed:
Finished test cases for getCandidateApplicationsByCandidateId() method

gabichelsea committed:
Created assertCandidateApplicationDTOCollection() method
Added willThrow() methods

kbertalan commented on service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java:
```diff
@@ -0,0 +1,335 @@
+    @Test(expected = IllegalArgumentException.class)
+    public void getStatesByApplicationIdShouldThrowIllegalArgumentExceptionWhenApplicationIdIsNull() {
+        // Given
+        given(applicationsRepository.findOne(null)).willThrow(IllegalArgumentException.class);
```
Do not call repository method, when you can detect if the parameter is wrong.
Use assert in the service implementation for catching invalid parameters.

kbertalan commented on service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java:
```diff
@@ -0,0 +1,335 @@
+    public void getStatesByApplicationIdShouldThrowIllegalArgumentExceptionWhenApplicationIsNull() {
+        // Given
+        given(applicationsRepository.findOne(FIRST_ID)).willReturn(null);
+        given(statesRepository.findByApplicationEntityOrderByStateIndexDesc(null)).willThrow(IllegalArgumentException.class);
```
It seems this given statement is not necessary anymore.

kbertalan commented on service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java:
```diff
@@ -0,0 +1,335 @@
+        assertThat(stateViewDTOList, notNullValue());
+        assertThat(stateViewDTOList.isEmpty(), is(true));
+
+        then(applicationsRepository).should(times(AT_ONE_TIME)).findOne(FIRST_ID);
```
should(times(1)) is the same as should()

kbertalan commented on service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java:
```diff
@@ -0,0 +1,335 @@
+    @Test(expected = IllegalArgumentException.class)
+    public void saveStateShouldThrowIllegalArgumentExceptionWhenApplicationIdIsNull() {
+        // Given
+        given(applicationsRepository.findOne(null)).willThrow(IllegalArgumentException.class);
```
it seems this given statement is not necessary

kbertalan commented on service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java:
```diff
@@ -0,0 +1,335 @@
+    @Test(expected = IllegalArgumentException.class)
+    public void saveStateShouldThrowIllegalArgumentExceptionWhenStateIsNull() {
+        // Given
+        given(statesRepository.save(stateEntity)).willThrow(IllegalArgumentException.class);
```
also this one seems unnecessary

kbertalan commented on service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java:
```diff
@@ -0,0 +1,335 @@
+        given(statesRepository.save(stateEntity)).willReturn(stateEntity);
+
+        // When
+        Long expectedId = statesService.saveState(stateDTO, FIRST_ID);
```
rename the variable as the returned is not the expected one

kbertalan commented on service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java:
```diff
@@ -0,0 +1,335 @@
+        assertThat(stateEntity, notNullValue());
+        assertThat(stateEntity.getCreationDate(), greaterThanOrEqualTo(now));
+        assertThat(stateEntity.getApplicationEntity(), equalTo(applicationEntity));
+        assertThat(stateEntity.getId(), equalTo(expectedId));
```
this assertion is not testing the logic executed in the service

kbertalan commented on service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java:
```diff
@@ -0,0 +1,335 @@
+    @Test(expected = IllegalArgumentException.class)
+    public void getCandidateApplicationsByCandidateIdShouldThrowIllegalArgumentExceptionWhenCandidateIdIsNull() {
+        // Given
+        given(candidateRepository.findOne(null)).willThrow(IllegalArgumentException.class);
```
remove this line

kbertalan commented on service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java:
```diff
@@ -0,0 +1,335 @@
+    public void getCandidateApplicationsByCandidateIdShouldThrowIllegalArgumentExceptionWhenCandidateEntityIsNull() {
+        // Given
+        given(candidateRepository.findOne(FIRST_ID)).willReturn(null);
+        given(applicationsRepository.findByCandidateEntity(null)).willThrow(IllegalArgumentException.class);
```
remove this line

kbertalan commented on service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java:
```diff
@@ -0,0 +1,335 @@
+        stateDTO = StateDTO.builder().id(FIRST_ID).build();
+
+        applicationEntityListWithSingleElement = Arrays.asList(applicationEntity);
+        applicationEntityListWithThreeElements = Arrays.asList(applicationEntity, applicationEntity, applicationEntity);
```
You should test with different application entities

kbertalan commented on service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java:
```diff
@@ -0,0 +1,335 @@
+        // Given
+        given(candidateRepository.findOne(FIRST_ID)).willReturn(candidateEntity);
+        given(applicationsRepository.findByCandidateEntity(candidateEntity)).willReturn(applicationEntityListWithThreeElements);
+        given(statesRepository.findTopByApplicationEntityOrderByStateIndexDesc(applicationEntity)).willReturn(stateEntity);
```
There should be 3 different applications in the returned list.

gabichelsea committed:
Removed unnecessary lines
Fixed it saveStateShouldBeSuccessSaving() method
Created different application entities

kbertalan commented on service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java:
```diff
@@ -263,50 +281,53 @@ public void getCandidateApplicationsByCandidateIdShouldReturnAnEmptyList() {
+        then(applicationsRepository).should().findByCandidateEntity(firstCandidateEntity);
+        then(statesRepository).should(times(TWO_TIMES)).findTopByApplicationEntityOrderByStateIndexDesc(firstApplicationEntity);
+        then(statesRepository).should(times(TWO_TIMES)).findTopByApplicationEntityOrderByStateIndexDesc(secondApplicationEntity);
+        then(statesRepository).should(times(TWO_TIMES)).findTopByApplicationEntityOrderByStateIndexDesc(thirdApplicationEntity);
```
The StateServiceImpl executes a single stateRepository.findTopByApplicationEntityOrderByStateIndexDesc() for each application entities.

Please remove the usage of the mock in the getExpectedCandidateApplicationDTOList.... method.

gabichelsea committed:
Compared Date objects fixed it.
Removed stateRepository.findTopByApplicationEntityOrderByStateIndexDesc() calling from the getExpectedCandidateApplicationDTOList.... method.

kbertalan commented on service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java:
```diff
@@ -337,22 +336,41 @@ private void assertCandidateApplicationDTOCollection(Collection<CandidateApplica
+        assertThat(stateEntity.getCreationDate(), greaterThanOrEqualTo(expectedSavedStateEntity.getCreationDate()));
+    }
+
+    private StateEntity getExpectedSavedStateEntity() {
```
I don't see your logic when to place a test data into a field initialized by the setup method or into a separate method, which is called from the test cases.

This comment is also apply for the next two methods.

gabichelsea committed:
Removed initialized methods
